contracts/StakingManager.sol

```diff
@@ -0,0 +1,166 @@
+     * Indexer's self delegations are not allow to redelegate.
+     */
+    function redelegate(
+        address from_indexer,
+        address to_indexer,
+        uint256 _amount
+    ) external override {
+        Staking staking = Staking(settings.getStaking());
+        address _source = msg.sender;
+        require(from_indexer != msg.sender, 'G004');
+        // delegation limit should not exceed
+        staking._checkDelegateLimitation(to_indexer, _amount);
+
+        staking.reflectEraUpdate(_source, from_indexer);
+        staking._removeDelegation(_source, from_indexer, _amount);
+        staking.reflectEraUpdate(_source, to_indexer);
+        staking._addDelegation(_source, to_indexer, _amount);
+    }
+
+    function cancelUnbonding(uint256 unbondReqId) external {
+        Staking staking = Staking(settings.getStaking());
+        require(unbondReqId >= staking.withdrawnLength(msg.sender), 'S007');
+        (address indexer, uint256 amount, uint256 startTime) = staking.unbondingAmount(msg.sender, unbondReqId);
+        require(amount > 0, 'S007');
+        IIndexerRegistry indexerRegistry = IIndexerRegistry(settings.getIndexerRegistry());
+        require(indexerRegistry.isIndexer(indexer), 'S007');
+
+        staking._removeUnbondingAmount(msg.sender, unbondReqId);
+        if (msg.sender != indexer) {
+            staking._checkDelegateLimitation(indexer, amount);
+        }
+        staking._addDelegation(msg.sender, indexer, amount);
+
+        emit UnbondCancelled(msg.sender, indexer, amount, unbondReqId);
+    }
+
+    /**
+     * @dev Withdraw max 10 mature unbond requests from an indexer.
+     * Each withdraw need to exceed lockPeriod.
+     */
+    function widthdraw() external override {
+        Staking staking = Staking(settings.getStaking());
+        require(!IDisputeManager(settings.getDisputeManager()).isOnDispute(msg.sender), 'G006');
+        uint256 withdrawingLength = staking.unbondingLength(msg.sender) - staking.withdrawnLength(msg.sender);
+        require(withdrawingLength > 0, 'S009');
+
+        uint256 latestWithdrawnLength = staking.withdrawnLength(msg.sender);
+        for (uint256 i = latestWithdrawnLength; i < latestWithdrawnLength + withdrawingLength; i++) {
+            (,,uint256 startTime) = staking.unbondingAmount(msg.sender, i);
+            if (block.timestamp - startTime < staking.lockPeriod()) {
+                break;
+            }
+
+            staking._withdrawARequest(msg.sender, i);
+        }
+    }
+
+    function slashIndexer(address _indexer, uint256 _amount) external {
+        require(msg.sender == settings.getDisputeManager(), 'G005');
+        Staking staking = Staking(settings.getStaking());
+        require(_amount <= staking.getSlashableAmount(_indexer), 'S010');
+        
+        staking._slashIndexer(_indexer, _amount);
+    }
+
+    function stakeCommission(address _indexer, uint256 _amount) external {
```

Comment:
The function to add commission fee to unbounding amount  can place inside `stakingContract`,